tools/nnyeah/nnyeah/Reworker.cs

```diff
@@ -0,0 +1,481 @@
+using System;
+using System.IO;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Cecil.Rocks;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace nnyeah {
+	public class Reworker {
+		Stream stm;
+		ModuleDefinition module;
+		TypeDefinition embeddedAttributeTypeDef;
+		TypeDefinition nativeIntegerAttributeTypeDef;
+		TypeReference nativeIntegerAttributeTypeRef;
+		TypeReference compilerGeneratedAttributeTypeRef;
+		TypeReference embeddedAttributeTypeRef;
+		TypeReference nintTypeReference;
+		TypeReference nuintTypeReference;
+
+		Dictionary<string, Transformation> methodSubs;
+		Dictionary<string, Transformation> fieldSubs;
+
+		public Reworker (Stream stm)
+		{
+			this.stm = stm;
+		}
+
+		public void Load ()
+		{
+			if (module != null)
+				return;
+			module = ModuleDefinition.ReadModule (stm);
+		}
+
+		public bool NeedsReworking ()
+		{
+			// simple prediate for seeing if there any references
+			// to types we need to care about.
+			// IntPtr is for future handling of types that
+			// descend from NObject and need to have the constructor
+			// changed to NHandle
+			return module.GetTypeReferences ().Any (
+				tr =>
+					tr.FullName == "System.nint" ||
+					tr.FullName == "System.nuint" ||
+					tr.FullName == "System.nfloat" ||
+					tr.FullName == "System.IntPtr"
+				);
+		}
+
+		void CheckModule ()
+		{
+			if (module == null)
+				throw new Exception ("Module is not loaded. Call Load first.");
```

Comment:
NRE exception instead?

Reply:
We shouldn't throw NRE ourselves. InvalidOperationException comes to mind, but I think just Exception is good enough.

Reply:
Nah. This would be a pilot error. I try to write the code such that it's an API and the workflow for API for this is:
- lightweight constructor
- `Load ()` (could be a candidate for being async)
- `if (NeedsWorking()) { Rework (outputstm); }` (also could be async)

In theory, the guts of `Load` could be moved into the ctor which would clean up a lot of the nullability assertions, but I **really** hate constructors that take forever and `Load` is not lightweight by any stretch of the imagination.

So if `CheckModule` gets called before `Load` it means the caller is not following the API and that's an error that should happen no more than once for any given project.

Reply:
I have added a comment on how to use the NoNullAttribute in the parameter to avoid the use !

Reply:
I prefer `InvalidOperationException` for "I screwed something up" error.